src/storage/src/hummock/block_cache.rs

```diff
@@ -12,56 +12,117 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
 use std::sync::Arc;
 
-use bytes::{BufMut, Bytes, BytesMut};
 use futures::Future;
-use moka::future::Cache;
 
-use super::{Block, HummockError, HummockResult, DEFAULT_ENTRY_SIZE};
+use super::cache::{CachableEntry, LookupResult, LruCache};
+use super::{Block, HummockError, HummockResult};
+
+const CACHE_SHARD_BITS: usize = 6; // It means that there will be 64 shards lru-cache to avoid lock conflict.
+const DEFAULT_OBJECT_POOL_SIZE: usize = 1024; // we only need a small object pool because when the cache reach the limit of capacity, it will
+                                              // always release some object after insert a new block.
+
+enum BlockEntry {
+    Cache(CachableEntry<(u64, u64), Box<Block>>),
+    Owned(Box<Block>),
+}
+
+pub struct BlockHolder {
+    _handle: BlockEntry,
+    block: *const Block,
+}
+
+impl BlockHolder {
+    pub fn from_owned_block(block: Box<Block>) -> Self {
+        let ptr = block.as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::Owned(block),
+            block: ptr,
+        }
+    }
+
+    pub fn from_cached_block(entry: CachableEntry<(u64, u64), Box<Block>>) -> Self {
+        let ptr = entry.value().as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::Cache(entry),
+            block: ptr,
+        }
+    }
+}
+
+impl AsRef<Block> for BlockHolder {
```

Comment:
May also want to implement `Deref<Block>` for `BlockHolder` so that we can avoid using `as_ref` everywhere.